### PR TITLE
Better Python login action ConnectionError exception handling

### PR DESCRIPTION
**Pull request by gurevichmark**
1. When `login` action fails, exception message contains some useful information that can be extracted and displayed to the user to help in debug process.
2. Formatting is also adjusted to match the Perl output 
    a. "BMC NotReady" is NOT displayed if login failure
    b. Login failure messages displayed to the console by default, not just for verbose
3. RC=1 is returned if login action fails.

```
[root@briggs01 xcat]# rpower mid05tor12cn05 bmcreboot
mid05tor12cn05: BMC reboot
[root@briggs01 xcat]# while true; do  date; rpower mid05tor12cn05 bmcstate; sleep 5; done
Wed Feb 28 15:58:38 EST 2018
mid05tor12cn05: Error: [500] Login to BMC failed: 500 Can't connect to 172.11.139.5 (Connection reset by peer).
Error: mid05tor12cn05: Error: BMC did not respond. Validate BMC configuration and retry the command.
Wed Feb 28 15:58:50 EST 2018
mid05tor12cn05: Error: [500] Login to BMC failed: 500 Can't connect to 172.11.139.5 (Connection reset by peer).
Error: mid05tor12cn05: Error: BMC did not respond. Validate BMC configuration and retry the command.
Wed Feb 28 15:58:59 EST 2018
mid05tor12cn05: Error: [500] Login to BMC failed: 500 Can't connect to 172.11.139.5:443 (Connection refused).
Error: mid05tor12cn05: Error: BMC did not respond. Validate BMC configuration and retry the command.
Wed Feb 28 15:59:05 EST 2018
mid05tor12cn05: Error: [500] Login to BMC failed: 500 Can't connect to 172.11.139.5:443 (timeout).
Error: mid05tor12cn05: Error: BMC did not respond. Validate BMC configuration and retry the command.
Wed Feb 28 15:59:40 EST 2018
mid05tor12cn05: Error: [500] Login to BMC failed: 500 Can't connect to 172.11.139.5:443 (No route to host).
Error: mid05tor12cn05: Error: BMC did not respond. Validate BMC configuration and retry the command.
Wed Feb 28 15:59:52 EST 2018
mid05tor12cn05: Error: [500] Login to BMC failed: 500 Can't connect to 172.11.139.5:443 (timeout).
Error: mid05tor12cn05: Error: BMC did not respond. Validate BMC configuration and retry the command.
Wed Feb 28 16:00:28 EST 2018
mid05tor12cn05: Error: [500] Login to BMC failed: 500 Can't connect to 172.11.139.5:443 (timeout).
Error: mid05tor12cn05: Error: BMC did not respond. Validate BMC configuration and retry the command.
Wed Feb 28 16:01:03 EST 2018
mid05tor12cn05: Error: [500] Login to BMC failed: 500 Can't connect to 172.11.139.5:443 (timeout).
Error: mid05tor12cn05: Error: BMC did not respond. Validate BMC configuration and retry the command.
Wed Feb 28 16:01:39 EST 2018
Error: mid05tor12cn05: Error: [404] org.freedesktop.DBus.Error.FileNotFound: path or object not found: /xyz/openbmc_project/state/bmc0
Wed Feb 28 16:01:52 EST 2018
mid05tor12cn05: BMC NotReady
Wed Feb 28 16:01:59 EST 2018
mid05tor12cn05: BMC NotReady
Wed Feb 28 16:02:08 EST 2018
mid05tor12cn05: BMC NotReady
Wed Feb 28 16:02:14 EST 2018
mid05tor12cn05: BMC NotReady
Wed Feb 28 16:02:20 EST 2018
mid05tor12cn05: BMC NotReady
Wed Feb 28 16:02:26 EST 2018
mid05tor12cn05: BMC NotReady
Wed Feb 28 16:02:31 EST 2018
mid05tor12cn05: BMC NotReady

```

### CODE REVIEW

**Comment by gurevichmark** [comment]
There is a small problem with lines like:
```
Error: mid05tor12cn05: Error: xxxxxx
```

This is because **self.messager.error** explained in #4854

File: xCAT-openbmc-py/lib/python/agent/hwctl/openbmc_client.py [hunk 1]
```diff
@@ -298,9 +307,13 @@ def request (self, method, resource, headers=None, payload=None, timeout=30, cmd
...
-            e.message = 'Error: BMC did not respond. ' \
-                        'Validate BMC configuration and retry the command.'
-            self._print_record_log(e.message, cmd)
+            message = 'Error: BMC did not respond. ' \
+                      'Validate BMC configuration and retry the command.'
+            if cmd == 'login':
+                login_message = "Error: [{0}] Login to BMC failed: {1} Can't connect to {2} {3}.".format(e.code, e.code, e.host_and_port, e.detail_msg)
+                self._print_record(login_message, cmd, "I")
+
+            self._print_record(message, cmd, "E")
```
Comment (xuweibj) [inline xCAT-openbmc-py/lib/python/agent/hwctl/openbmc_client.py]: I want to make sure that does this mean that if the request failed will print out error msg for all commands? I think may have command need to handle this type error itself, I think print here is not a good idea.

File: xCAT-openbmc-py/lib/python/agent/hwctl/openbmc_client.py [hunk 2]
```diff
@@ -276,7 +285,7 @@ def handle_response (self, resp, cmd=''):
         if code != requests.codes.ok:
             description = ''.join(data['data']['description'])
             error = 'Error: [%d] %s' % (code, description)
-            self._print_record_log(error, cmd)
+            self._print_record(error, cmd, "E")
```
Comment (xuweibj) [inline xCAT-openbmc-py/lib/python/agent/hwctl/openbmc_client.py]: Several requests need to handle error themselves, I think we should not print out here.

**Comment by xuweibj** [comment]
Hi @gurevichmark , should we print out msg the same with perl code?
```
[500] Login to BMC failed: 500 Can't connect to 172.11.139.5 (Connection reset by peer).
```

In perl, we actually get the string "Can't connect to 172.11.139.5" from LWP, we in python, we did not. I think we could know the exact status by``BMC did not respond. Validate BMC configuration and retry the command.``.

**Comment by gurevichmark** [comment]
Have a dependency on #4871 to properly display error message

**Comment by gurevichmark** [comment]
@xuweibj Victor and I talked about this PR. We decided:
1. Not to display "fake" 500 code
2. Display `ip:port` information, helpful in debugging
3. Display only one line of output about login failure

Output is now:
```
[root@briggs01 xcat]# export XCAT_OPENBMC_PYTHON=YES
[root@briggs01 xcat]# rpower mid05tor12cn[05,13] bmcreboot
mid05tor12cn13: BMC reboot
mid05tor12cn05: BMC reboot
[root@briggs01 xcat]# while true; do  date; rpower p9up bmcstate; sleep 5; done
Fri Mar  2 10:49:29 EST 2018
mid05tor12cn16: BMC Ready
mid05tor12cn15: BMC Ready
mid05tor12cn02: Error: Login to BMC failed: Can't connect to 172.11.139.2:443 (No route to host).
mid05tor12cn13: Error: Login to BMC failed: Can't connect to 172.11.139.13 (Connection reset by peer).
mid05tor12cn05: Error: Login to BMC failed: Can't connect to 172.11.139.5 (Connection reset by peer).
Fri Mar  2 10:49:41 EST 2018
mid05tor12cn13: Error: Login to BMC failed: Can't connect to 172.11.139.13:443 (Connection refused).
mid05tor12cn05: Error: Login to BMC failed: Can't connect to 172.11.139.5:443 (Connection refused).
mid05tor12cn02: Error: Login to BMC failed: Can't connect to 172.11.139.2:443 (No route to host).
mid05tor12cn15: BMC Ready
mid05tor12cn16: BMC Ready
Fri Mar  2 10:49:47 EST 2018
mid05tor12cn15: BMC Ready
mid05tor12cn16: BMC Ready
mid05tor12cn02: Error: Login to BMC failed: Can't connect to 172.11.139.2:443 (No route to host).
mid05tor12cn13: Error: Login to BMC failed: Can't connect to 172.11.139.13:443 (timeout).
mid05tor12cn05: Error: Login to BMC failed: Can't connect to 172.11.139.5:443 (timeout).
Fri Mar  2 10:50:23 EST 2018
mid05tor12cn02: Error: Login to BMC failed: Can't connect to 172.11.139.2:443 (No route to host).
mid05tor12cn16: BMC Ready
mid05tor12cn15: BMC Ready
mid05tor12cn05: Error: Login to BMC failed: Can't connect to 172.11.139.5:443 (No route to host).
mid05tor12cn13: Error: Login to BMC failed: Can't connect to 172.11.139.13:443 (No route to host).
Fri Mar  2 10:50:30 EST 2018
mid05tor12cn16: BMC Ready
mid05tor12cn15: BMC Ready
mid05tor12cn05: Error: Login to BMC failed: Can't connect to 172.11.139.5:443 (No route to host).
mid05tor12cn02: Error: Login to BMC failed: Can't connect to 172.11.139.2:443 (No route to host).
mid05tor12cn13: Error: Login to BMC failed: Can't connect to 172.11.139.13:443 (No route to host).
Fri Mar  2 10:50:38 EST 2018
mid05tor12cn15: BMC Ready
mid05tor12cn16: BMC Ready
mid05tor12cn13: Error: Login to BMC failed: Can't connect to 172.11.139.13:443 (No route to host).
mid05tor12cn05: Error: Login to BMC failed: Can't connect to 172.11.139.5:443 (No route to host).
mid05tor12cn02: Error: Login to BMC failed: Can't connect to 172.11.139.2:443 (No route to host).
Fri Mar  2 10:50:47 EST 2018
mid05tor12cn15: BMC Ready
mid05tor12cn16: BMC Ready
mid05tor12cn02: Error: Login to BMC failed: Can't connect to 172.11.139.2:443 (No route to host).
mid05tor12cn13: Error: Login to BMC failed: Can't connect to 172.11.139.13:443 (No route to host).
mid05tor12cn05: Error: Login to BMC failed: Can't connect to 172.11.139.5:443 (timeout).
Fri Mar  2 10:51:22 EST 2018
mid05tor12cn16: BMC Ready
mid05tor12cn15: BMC Ready
mid05tor12cn02: Error: Login to BMC failed: Can't connect to 172.11.139.2:443 (No route to host).
mid05tor12cn13: Error: Login to BMC failed: Can't connect to 172.11.139.13:443 (timeout).
mid05tor12cn05: Error: Login to BMC failed: Can't connect to 172.11.139.5:443 (timeout).
Fri Mar  2 10:51:58 EST 2018
mid05tor12cn16: BMC Ready
mid05tor12cn15: BMC Ready
mid05tor12cn02: Error: Login to BMC failed: Can't connect to 172.11.139.2:443 (No route to host).
mid05tor12cn13: BMC NotReady
mid05tor12cn05: Error: Login to BMC failed: Can't connect to 172.11.139.5:443 (timeout).
Fri Mar  2 10:52:34 EST 2018
mid05tor12cn15: BMC Ready
mid05tor12cn16: BMC Ready
mid05tor12cn13: BMC NotReady
mid05tor12cn02: Error: Login to BMC failed: Can't connect to 172.11.139.2:443 (No route to host).
mid05tor12cn05: BMC NotReady
Fri Mar  2 10:52:46 EST 2018
mid05tor12cn15: BMC Ready
mid05tor12cn16: BMC Ready
mid05tor12cn05: BMC NotReady
mid05tor12cn02: Error: Login to BMC failed: Can't connect to 172.11.139.2:443 (No route to host).
mid05tor12cn13: BMC Ready
Fri Mar  2 10:52:56 EST 2018
mid05tor12cn15: BMC Ready
mid05tor12cn13: BMC Ready
mid05tor12cn16: BMC Ready
mid05tor12cn05: BMC Ready
mid05tor12cn02: Error: Login to BMC failed: Can't connect to 172.11.139.2:443 (No route to host).
[root@briggs01 xcat]#
```

**Comment by xuweibj** [comment]
Looks good to me. Agree to merge.

**Pull request merged**